File: js/movement/captureQuality.js
/* ─── Movement Assessment — capture (measurement) quality ─────────────────
 *
 * Describes how trustworthy the MEASUREMENT is. It says nothing about the
 * movement itself: "sufficient" does not mean good movement and
 * "insufficient" does not mean bad movement.
 *
 *   sufficient    a repetition was segmented and every factor passed
 *   limited       a repetition was segmented but at least one factor is
 *                 degraded; the numbers are shown with that caveat
 *   insufficient  no repetition could be segmented, or too little of the
 *                 capture was usable; no metrics are shown
 *
 * Factors (measurement parameters, documented in MOVEMENT_ASSESSMENT.md):
 *   usable_frames      share of capture frames with a valid knee angle
 *                      (limited < 0.8, insufficient < 0.5)
 *   landmark_confidence median, over valid samples, of the lowest provider
 *                      visibility among hip/knee/ankle (limited < 0.75)
 *   single_person      frames where the provider returned two poses
 *                      (insufficient if any: subject identity is ambiguous)
 *   body_in_frame      share of pose frames with head and the analysed
 *                      leg/foot in the image (limited < 0.9)
 *   foot_stability     90th-percentile displacement of the analysed ankle
 *                      from its calibration position, divided by apparent
 *                      standing height in pixels (limited > 0.08). Feet stay
 *                      planted in a bodyweight squat, so movement here means
 *                      the camera or the feet moved; the two cannot be told
 *                      apart from one camera.
 *   repetition         whether one repetition was segmented
 * ───────────────────────────────────────────────────────────────────────── */

import { ANGLE_DEFINITIONS, angleSample } from "./kinematics.js";
import { fullBodyInFrame } from "./calibration.js";
import { distance, median, percentile } from "./geometry.js";
import { sideLandmarkName } from "./poseContract.js";

export const QUALITY = Object.freeze({
  minUsableFraction: 0.8,
  insufficientUsableFraction: 0.5,
  minMedianVisibility: 0.75,
  minBodyInFrameFraction: 0.9,
  maxFootDriftFraction: 0.08,
  footDriftPercentile: 90
});

function round(value, digits) {
  if (typeof value !== "number" || !Number.isFinite(value)) return null;
  const f = Math.pow(10, digits);
  return Math.round(value * f) / f;
}

export function assessCaptureQuality(input, options) {
  const opt = Object.assign({}, QUALITY, options || {});
  const frames = (input && input.captureFrames) || [];
  const side = input && input.side;
  const calibration = (input && input.calibration) || {};
  const segmentation = (input && input.segmentation) || { state: "insufficient" };

  const totalFrames = frames.length;
  const stamps = frames.map(function (f) { return f && f.timestampMs; })
    .filter(function (t) { return typeof t === "number" && Number.isFinite(t); });
  const captureDurationMs = stamps.length > 1 ? stamps[stamps.length - 1] - stamps[0] : 0;
  const counts = { pose: 0, no_pose: 0, multiple_poses: 0, malformed: 0 };
  frames.forEach(function (f) {
    const s = f && f.status;
    counts[s in counts ? s : "malformed"] += 1;
  });

  const knee = side ? frames.map(function (f) { return angleSample(f, side, ANGLE_DEFINITIONS.knee); }) : [];
  const validKnee = knee.filter(function (s) { return s.state === "valid"; });
  const usableFraction = totalFrames > 0 ? validKnee.length / totalFrames : 0;

  const visibilities = [];
  const ankleName = side ? sideLandmarkName(side, "ankle") : null;
  frames.forEach(function (f, i) {
    if (!knee[i] || knee[i].state !== "valid") return;
    const lm = f.landmarks;
    visibilities.push(Math.min(
      lm[sideLandmarkName(side, "hip")].visibility,
      lm[sideLandmarkName(side, "knee")].visibility,
      lm[ankleName].visibility
    ));
  });
  const medianVisibility = median(visibilities);

  const poseFrames = frames.filter(function (f) { return f && f.status === "pose"; });
  const inFrame = side ? poseFrames.filter(function (f) { return fullBodyInFrame(f, side); }).length : 0;
  const bodyInFrameFraction = poseFrames.length > 0 ? inFrame / poseFrames.length : 0;

  let footDrift = null;
  const ref = calibration.ankleReference;
  if (side && ref && typeof calibration.standingHeightPx === "number" && calibration.standingHeightPx > 0) {
    const drifts = poseFrames
      .map(function (f) { return f.landmarks[ankleName]; })
      .filter(function (lm) { return lm && lm.inFrame; })
      .map(function (lm) { return distance(lm, ref); });
    const p = percentile(drifts, opt.footDriftPercentile);
    footDrift = p === null ? null : p / calibration.standingHeightPx;
  }

  const segmented = segmentation.state === "segmented";
  const factors = [
    {
      id: "usable_frames",
      state: usableFraction < opt.insufficientUsableFraction ? "insufficient" : usableFraction < opt.minUsableFraction ? "limited" : "pass",
      value: round(usableFraction, 3)
    },
    {
      id: "landmark_confidence",
      state: medianVisibility === null ? "insufficient" : medianVisibility < opt.minMedianVisibility ? "limited" : "pass",
      value: round(medianVisibility, 3)
    },
    {
      id: "single_person",
      state: counts.multiple_poses > 0 ? "insufficient" : "pass",
      value: counts.multiple_poses
    },
    {
      id: "body_in_frame",
      state: bodyInFrameFraction < opt.minBodyInFrameFraction ? "limited" : "pass",
      value: round(bodyInFrameFraction, 3)
    },
    {
      id: "foot_stability",
      state: footDrift === null ? "limited" : footDrift > opt.maxFootDriftFraction ? "limited" : "pass",
      value: round(footDrift, 3)
    },
    {
      id: "repetition",
      state: segmented ? "pass" : "insufficient",
      value: segmented ? "one_repetition" : (segmentation.reason || "not_segmented")
    }
  ];

  let state = "sufficient";
  if (factors.some(function (f) { return f.state === "insufficient"; })) state = "insufficient";
  else if (factors.some(function (f) { return f.state === "limited"; })) state = "limited";

  return {
    state: state,
    factors: factors,
    totalFrames: totalFrames,
    captureDurationMs: round(captureDurationMs, 0),
    usableFrames: validKnee.length,
    usableFraction: round(usableFraction, 3),
    medianLandmarkVisibility: round(medianVisibility, 3),
    frameStatusCounts: counts,
    bodyInFrameFraction: round(bodyInFrameFraction, 3),
    footDriftFraction: round(footDrift, 3)
  };
}

File: research_backend/physiq_research/domain/capture_quality.py
"""Capture (measurement) quality (port of captureQuality.js).

Describes how trustworthy the MEASUREMENT is — never the movement.
"sufficient" does not mean good movement and "insufficient" does not mean
bad movement.

    usable_frames        share of capture frames with a valid knee angle
                         (limited < 0.8, insufficient < 0.5)
    landmark_confidence  median over valid samples of the lowest
                         hip/knee/ankle visibility (limited < 0.75)
    single_person        frames with two poses (insufficient if any)
    body_in_frame        share of pose frames with head and analysed
                         leg/foot in view (limited < 0.9)
    foot_stability       90th-percentile ankle displacement from its
                         calibration position ÷ apparent standing height
                         (limited > 0.08; camera or feet moved)
    repetition           whether one repetition was segmented
"""

from __future__ import annotations

import math
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any, Final

from physiq_research.domain.calibration import Calibration, full_body_in_frame
from physiq_research.domain.geometry import distance, median, percentile
from physiq_research.domain.jsnum import js_round_digits
from physiq_research.domain.kinematics import KNEE, angle_sample
from physiq_research.domain.pose_frame import PoseFrame, side_landmark_name
from physiq_research.domain.segmentation import Segmentation

QUALITY: Final[dict[str, Any]] = {
    "minUsableFraction": 0.8,
    "insufficientUsableFraction": 0.5,
    "minMedianVisibility": 0.75,
    "minBodyInFrameFraction": 0.9,
    "maxFootDriftFraction": 0.08,
    "footDriftPercentile": 90,
}


@dataclass(frozen=True, slots=True)
class QualityFactor:
    id: str
    state: str  # "pass" | "limited" | "insufficient"
    value: float | int | str | None  # unrounded (research precision)


@dataclass(frozen=True, slots=True)
class CaptureQuality:
    state: str  # "sufficient" | "limited" | "insufficient"
    factors: tuple[QualityFactor, ...]
    total_frames: int
    capture_duration_ms: float
    usable_frames: int
    usable_fraction: float
    median_landmark_visibility: float | None
    frame_status_counts: dict[str, int]
    body_in_frame_fraction: float
    foot_drift_fraction: float | None

    def m6_view(self, *, rounded: bool = True) -> dict[str, Any]:
        """The values rounded exactly as movement-assessment-v0.1 reports them
        (or unrounded, same structure, with ``rounded=False``)."""

        def r(v: Any, d: int) -> Any:
            if not rounded:
                return v
            return js_round_digits(v, d, no_negative_zero=False) if isinstance(v, float) else v

        digits = {"usable_frames": 3, "landmark_confidence": 3, "body_in_frame": 3, "foot_stability": 3}
        return {
            "state": self.state,
            "factors": [
                {"id": f.id, "state": f.state, "value": r(f.value, digits[f.id]) if f.id in digits else f.value}
                for f in self.factors
            ],
            "totalFrames": self.total_frames,
            "captureDurationMs": r(self.capture_duration_ms, 0),
            "usableFrames": self.usable_frames,
            "usableFraction": r(self.usable_fraction, 3),
            "medianLandmarkVisibility": r(self.median_landmark_visibility, 3),
            "frameStatusCounts": dict(self.frame_status_counts),
            "bodyInFrameFraction": r(self.body_in_frame_fraction, 3),
            "footDriftFraction": r(self.foot_drift_fraction, 3),
        }


def assess_capture_quality(
    capture_frames: Sequence[PoseFrame],
    side: str | None,
    calibration: Calibration | None,
    segmentation: Segmentation,
    options: dict[str, Any] | None = None,
) -> CaptureQuality:
    opt = {**QUALITY, **(options or {})}
    frames = list(capture_frames)
    total = len(frames)
    stamps = [f.t_ms for f in frames if f is not None and f.t_ms is not None and math.isfinite(f.t_ms)]
    capture_duration = float(stamps[-1] - stamps[0]) if len(stamps) > 1 else 0.0
    counts = {"pose": 0, "no_pose": 0, "multiple_poses": 0, "malformed": 0}
    for f in frames:
        s = f.status if f is not None else None
        counts[s if s in counts else "malformed"] += 1

    knee = [angle_sample(f, side, KNEE) for f in frames] if side else []
    valid_knee = [s for s in knee if s.state == "valid"]
    usable_fraction = len(valid_knee) / total if total > 0 else 0.0

    visibilities: list[float] = []
    ankle_name = side_landmark_name(side, "ankle") if side else None
    if side:
        for i, f in enumerate(frames):
            if knee[i].state != "valid":
                continue
            lm = f.landmarks
            hip = lm[side_landmark_name(side, "hip")]
            kn = lm[side_landmark_name(side, "knee")]
            an = lm[ankle_name]  # type: ignore[index]
            assert hip is not None and kn is not None and an is not None
            visibilities.append(min(hip.visibility, kn.visibility, an.visibility))
    median_visibility = median(visibilities)

    pose_frames = [f for f in frames if f is not None and f.status == "pose"]
    in_frame = sum(1 for f in pose_frames if full_body_in_frame(f, side)) if side else 0
    body_in_frame_fraction = in_frame / len(pose_frames) if pose_frames else 0.0

    foot_drift: float | None = None
    ref = calibration.ankle_reference if calibration is not None else None
    height = calibration.standing_height_px if calibration is not None else None
    if side and ref is not None and isinstance(height, float) and height > 0:
        drifts = []
        for f in pose_frames:
            ankle = f.landmarks.get(ankle_name) if ankle_name else None
            if ankle is not None and ankle.in_frame:
                drifts.append(distance(ankle, ref))
        p = percentile(drifts, opt["footDriftPercentile"])
        foot_drift = None if p is None else p / height

    segmented = segmentation.state == "segmented"

    def usable_state(v: float) -> str:
        if v < opt["insufficientUsableFraction"]:
            return "insufficient"
        return "limited" if v < opt["minUsableFraction"] else "pass"

    factors = (
        QualityFactor("usable_frames", usable_state(usable_fraction), usable_fraction),
        QualityFactor(
            "landmark_confidence",
            "insufficient"
            if median_visibility is None
            else ("limited" if median_visibility < opt["minMedianVisibility"] else "pass"),
            median_visibility,
        ),
        QualityFactor(
            "single_person", "insufficient" if counts["multiple_poses"] > 0 else "pass", counts["multiple_poses"]
        ),
        QualityFactor(
            "body_in_frame",
            "limited" if body_in_frame_fraction < opt["minBodyInFrameFraction"] else "pass",
            body_in_frame_fraction,
        ),
        QualityFactor(
            "foot_stability",
            "limited" if foot_drift is None or foot_drift > opt["maxFootDriftFraction"] else "pass",
            foot_drift,
        ),
        QualityFactor(
            "repetition",
            "pass" if segmented else "insufficient",
            "one_repetition" if segmented else (segmentation.reason or "not_segmented"),
        ),
    )
    state = "sufficient"
    if any(f.state == "insufficient" for f in factors):
        state = "insufficient"
    elif any(f.state == "limited" for f in factors):
        state = "limited"

    return CaptureQuality(
        state=state,
        factors=factors,
        total_frames=total,
        capture_duration_ms=capture_duration,
        usable_frames=len(valid_knee),
        usable_fraction=usable_fraction,
        median_landmark_visibility=median_visibility,
        frame_status_counts=counts,
        body_in_frame_fraction=body_in_frame_fraction,
        foot_drift_fraction=foot_drift,
    )
